### ISSUE TITLE
PATCHing an EVSE via the URL of its location object

### ISSUE THREAD

**Issue by ahzf** [ocpi/ocpi#285]
I have some discussions with someone sending me locations, EVSEs, etc.pp

In generals OCPI understands:
- PATCH    ~/locations/{country_code}/{party_id}/{locationId}
- PATCH    ~/locations/{country_code}/{party_id}/{locationId}/{evseId}
- PATCH    ~/locations/{country_code}/{party_id}/{locationId}/{evseId}/{connectorId}

When someone now wants to patch the EVSE status (s)he could use the 2nd URL and everything is clear.

But some read the specification in a way, that PATCHing the EVSE status via the first (location) URL should also be possible. Here it starts to become a bit weird. Because how to adress the correct EVSE within the **evses** JSON property? What is the outcome of the following merge-PATCH document when a) the EVSE existed, or not existed before?

```
{
    "evses": [
        {
            "uid": "1234",
            "status": "AVAILABLE"
        }
    ]
}
```

- If the EVSE existed before, we could use the **uid** property to find the correct EVSE and do the PATCHing. But this is more than a simple Merge-Patch as defined in e.g. https://tools.ietf.org/html/rfc7386 . We have to implement a special handling of the **evses** property and tell it to use the **uid** property as lookup and all other properties for the real PATCHing (exccept the **connectors** which is again special). In the end we do what we could do more easily via using the EVSE PATCH URL (  `~/locations/{country_code}/{party_id}/{locationId}/{evseId}` ).
- If it did not existes before we could create a new EVSE. But in this example the **connectos** property is missing, so it is not a valid EVSE object. Anyway we have the EVSE PUT URL for exactly this use case.

So for both cases we could do the same much more easily using other methods. Should the OCPI specification for this reason not recommend to disallow any PATCHing of the **evses** property of locations and the **connectors** property of EVSEs? Saying that, I think it would also be good to disallow any PATCH for the **country_code**, **party_id**, **id** and **uid** properties of any object, Also setting e.g. the **evses**/**connectors** property to `null` or `[]` should be illegal... for the same reasonsI

So to make a simple rule: _Use the (Merge) Patch only for properties of objects, but not for identifiers and not for links to other objects!_ I would guess that you created the EVSEs and Connectors URLs for exactly this reason, but without a clear statement within the specification document some people come to weird conclusions about the intended protocol behaviour ;)

**Comment by poohsen** [comment]
if I had but one vote I'd vote for the PATCH to be removed from the protocol...
My proposal for how to do it is here: https://github.com/ocpi/ocpi/issues/146

For now I'd try to convince a partner to see the difficulty you outline and only use the PATCH in their implementation for changing the EVSE status.

**Comment by ahzf** [comment]
Well... PUT is also no free lunch. When you do HTTP in a correct way, you would do a **conditional PUT**, e.g. using E-Tags and "If-Match"-HTTP-Headers or the LastModifiedDate and "If-Unmodified-Since"-HTTP-Headers. But for this you also need a GET request before you can update a resource via PUT... or the strong trust that you a the only writer and your entire application is single-threaded, like back in the good old 1990ies ;)

PATCH is just the little cousin of PUT... as long as you limit the scope of a patch. If you are allowed to patch the whole data set with one request you probably have a really big problem. I think PATCH is the more polite way to do updates, as you explicitly say what you want. The receiver has to do a diff anyway, as (s)he might be interested in what you have updated or (s)he might only to store or log your changes.

PATCHing _the_ EVSE status is perhaps the biggest design flaw of OCPI. I really do not know anybody who does not filter out those requests and implements a different way to process them. I really thought it is common sense in e-mobility to separate static and dynamic data.

**Comment by poohsen** [comment]
> PATCHing the EVSE status is perhaps the biggest design flaw of OCPI
Agreed. That's why I propose a separate endpoint for changing the EVSE status. That would give an unambiguous way to change this one piece of dynamic data that we have in the locations module.

**Comment by ahzf** [comment]
Well... there is perhaps more than this "one piece" of dynamic EVSE status. Maybe you want a dynamic status for charging stations (which do not exist in OCPI yet, I know), charging locations, backends, ...

So perhaps it should be an entire OCPI module for dynamic data updates. What do you think? ;)

**Comment by simonschllng** [comment]
@ahzf you turn it back to the origins where we discussed whether we should separate dynamic from 'static' data (like in OCHP) or put all into one object. But other than OCHP this could really be more generic and not specifically on the charge point level. Let's se it as a message queue and let's exchange status messages. Sending some kind of "system heartbeat" from the backend could also add additional value to monitor the health of each other and the connection.

**Comment by simonschllng** [comment]
> > PATCHing the EVSE status is perhaps the biggest design flaw of OCPI
> 
> Agreed. That's why I propose a separate endpoint for changing the EVSE status. That would give an unambiguous way to change this one piece of dynamic data that we have in the locations module.

@poohsen do you recall what the arguments were to define it like that?

**Comment by poohsen** [comment]
IIRC, it was inspired by the notion of RESTfulness.

if you assume that 
1) both static and dynamic data are going to be put in the same resource (in this case: location) and
2) you don't want to send an EVSE object every time some nested value inside it changes 

then PATCH is a straightforward way to achieve this.

**Comment by ahzf** [comment]
Just ask e-clearing.net. As far as I understood they also implemented a special handling of the EVSE status.
